ffi/cpp/include/provider.hpp

```diff
@@ -0,0 +1,49 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+#ifndef ICU4X_PROVIDER_HPP
+#define ICU4X_PROVIDER_HPP
+
+#include <algorithm>
+#include <optional>
+#include <string_view>
+
+#include "../../capi/include/provider.h"
+
+namespace icu4x {
+class DataProvider {
+ public:
+  ~DataProvider() {
```

Comment:
Question: Why aren't you doing DataProvider the same way as the other classes, with a unique_ptr field?

Reply:
because it's a rust fat pointer, which in FFI is just represented as a struct of two usizes

unique_ptr is customizable so it is possible to use it here, but it requires some tricky stuff, some of which I'm not sure applies to these structs defined in C, and really doesn't seem worth it for just one type

Reply:
Okay.  I still don't really know what this is doing, and why you only ever read `_field1` and never `_field2`, so I would prefer if you would either:

1. (preferred) Implement [NullablePointer](https://en.cppreference.com/w/cpp/named_req/NullablePointer) for this type so that it can be used directly in std::unique_ptr
2. Add more documentation around ICU4XDataProvider, both here and in provider.h, explaining what a Rust fat pointer is and what the fields do and what the semantics are

I prefer option 1 in large part because I like to avoid custom implementations of copy/move/destruct when possible because they're hard to reason about.

Reply:
NullablePointer is the "tricky stuff" I was referencing: Our type is not equality comparable to bool, and it's defined in C, so we'd need to first wrap the type and create a type with a lot of additional things on it, at which point we're back in the same spot.

The semantics of a Rust fat pointer are largely supposed to be opaque, I don't really know what I _can_ write here. I'm relying on the fact that neither field is null on the Rust side and just nulling one out.